streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -220,11 +222,26 @@ public void setStateListener(final KafkaStreams.StateListener listener) {
         stateListener = listener;
     }
 
-    private void setState(final State newState) {
+    /**
+     * Sets the state
+     * @param newState New state
+     * @param ignoreWhenShuttingDownOrNotRunning, if true, then we'll first check if the state is
+     *                                            PENDING_SHUTDOWN or NOT_RUNNING, and if it is,
+     *                                            we immediately return. Effectively this enables
+     *                                            a conditional set, under the stateLock lock.
+     */
+    private void setState(final State newState, boolean ignoreWhenShuttingDownOrNotRunning) {
         synchronized (stateLock) {
+            if (ignoreWhenShuttingDownOrNotRunning) {
```

Comment:
@guozhangwang yes that seems correct. It would seem to be a bug if `setState` is called when were are in `NOT_RUNNING` state